### PR TITLE
MINOR: JavaDoc improvements for new state store API

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @bbejeck @dguy @guozhangwang @miguno

File: streams/src/main/java/org/apache/kafka/streams/Consumed.java [hunk 1]
```diff
@@ -24,10 +24,29 @@
 
 /**
  * The {@code Consumed} class is used to define the optional parameters when using {@link StreamsBuilder} to
- * build instancs of {@link KStream}, {@link KTable}, and {@link GlobalKTable}.
+ * build instances of {@link KStream}, {@link KTable}, and {@link GlobalKTable}.
+ * <p>
+ * For example, you can read an topic as {@link KStream} with custom timestamp extractor and specify the corresponding
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/Consumed.java]: Typos:
> You can read [a] topic as `KStream` with [a] custom timestamp extractor ...

File: streams/src/main/java/org/apache/kafka/streams/Consumed.java [hunk 2]
```diff
@@ -24,10 +24,29 @@
...
+ * key and value serdes like:
+ * <pre>{@code
+ * StreamsBuilder builder = new StreamsBuilder();
+ * KStream<String, Long> stream = builder.stream(
+ *   "topicName",
+ *   Consumed.with(Serdes.String(), Serdes.Long())
+ *           .withTimestampExtractor(new LogAndSkipOnInvalidTimestamp()));
+ * }</pre>
+ * Similar, you can read a topic as {@link KTable} with custom "auto.offset.reset" strategy and force a state store
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/Consumed.java]: Typo:
> Similar[ly], you can ... with [a] custom ...

File: streams/src/main/java/org/apache/kafka/streams/Consumed.java [hunk 3]
```diff
@@ -24,10 +24,29 @@
...
+ * {@link org.apache.kafka.streams.kstream.Materialized materialization} to access the content via Interactive Queries API:
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/Consumed.java]: Suggestion: We already say "Kafka Streams API", "DSL", and "Processor API".  That's already a lot of "APIs" even though it's all about the same thing.  I wouldn't label interactive queries to be yet another API.

I'd just say:
> ... to access the content via interactive queries:

File: streams/src/main/java/org/apache/kafka/streams/Consumed.java (hunk 2 above)
Comment (miguno) [reply streams/src/main/java/org/apache/kafka/streams/Consumed.java]: Consistency: Here we say

> with [a] custom "auto.offset.reset" strategy

and below we say 

> Enum used to define auto offset reset policy when creating {@link KStream} or {@link KTable}.

The AK docs say neither "strategy" or "policy", it's just a config like any other.  Hence I'd suggest to change the above consistently to sth like:

> with [a] custom {@code auto.offset.reset} configuration

and

> Sets the {@code auto.offset.reset} configuration when creating {@link KStream} or {@link KTable}.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java [hunk 4]
```diff
@@ -34,9 +34,23 @@
 
 /**
  * Used to describe how a {@link StateStore} should be materialized.
- * You can either provide a custom {@link StateStore} backend
- * through one of the provided methods accepting a supplier or use the default RocksDB backends
- * by providing just a store name.
+ * You can either provide a custom {@link StateStore} backend through one of the provided methods accepting a supplier
+ * or use the default RocksDB backends by providing just a store name.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java:38]: Typo:
> or use the default RocksDB backend[] by providing ...

File: streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java [hunk 5]
```diff
@@ -34,9 +34,23 @@
...
+ * <pre>{@code
+ * StreamsBuilder builder = new StreamsBuilder();
+ * KTable<Integer, Integer> table = builder.table(
+ *   "topicName",
+ *   Materialized.as("queryable-store-name"));
+ * }</pre>
+ *
+ * @param <K> type of record key
+ * @param <V> type of record value
+ * @param <S> type of state store (note, state stores always have key/value types {@code <Bytes,byte[]>}
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java]: Style nit for "note" and whitespace between `Bytes` and `byte[]`:

> (note, state stores always have key/value types {@code <Bytes,byte[]>}

should be

> (note: state stores always have key/value types {@code <Bytes, byte[]>}

Comment (miguno) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java]: Question: If the state stores are always of type `<Bytes, byte[]>`, why does the user even have to specify the type explicitly?

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 6]
```diff
@@ -44,6 +44,40 @@
 
 /**
  * Factory for creating state stores in Kafka Streams.
+ * <p>
+ * For high-level DSL, i.e., {@link org.apache.kafka.streams.StreamsBuilder StreamsBuilder}, users create
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: > For high-level DSL,

should be

> When using the high-level DSL, ...

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 7]
```diff
@@ -44,6 +44,40 @@
...
+ * StreamsBuilder builder = new StreamsBuilder();
+ * KeyValueBytesStoreSupplier storeSupplier = Stores.inMemoryKeyValueStore("queryable-store-name");
+ * KTable<Long,String> table = builder.table(
+ *   "topicName",
+ *   Materialized.as(storeSupplier)
+ *               .withKeySerde(Serdes.Long())
+ *               .withValueSerde(Serdes.String())
+ *               .withCachingDisabled());
+ * }</pre>
+ * For low-level Processor API, i.e., {@link org.apache.kafka.streams.Topology Topology}, users create
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: > For low-level Processor API,

should be

> When using the Processor API, ...

(IMHO we should also stop saying "low-level" PAPI.  It's simply a different API.)

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 8]
```diff
@@ -44,6 +44,40 @@
...
+ * {@link StoreBuilder}s that can be attached to {@link org.apache.kafka.streams.processor.Processor Processor}s.
+ * For example, you can create an {@link org.apache.kafka.streams.kstream.Windowed windowed} RocksDB store with custom
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: Typo: "you can create [a] windowed ..."

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 9]
```diff
@@ -44,6 +44,40 @@
...
+ * changelog configuration like:
+ * <pre>{@code
+ * Topology topology = new Topology();
+ * topology.addProcessor("processorName", ...);
+ *
+ * Map<String,String> changelogConfig = new HashMap<>();
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: Suggestion:  I am not sure which terminology we use elsewhere, but "changelogConfig" may suggest to the average reader that there's a separate "changelog" concept that they may have never heard of.  It's simply a topic configuration.

I'd suggest to rename `changelogConfig` to `changelogTopicConfig` or the shorter `topicConfig`.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 10]
```diff
@@ -44,6 +44,40 @@
...
  * Factory for creating state stores in Kafka Streams.
+ * <p>
+ * For high-level DSL, i.e., {@link org.apache.kafka.streams.StreamsBuilder StreamsBuilder}, users create
+ * {@link StoreSupplier}s that can be further customized via
+ * {@link org.apache.kafka.streams.kstream.Materialized Materialized}.
+ * For example, a topic read as {@link org.apache.kafka.streams.kstream.KTable KTable} can be materialized into an
+ * in-memory store with custom key/value serdes and caching disabled:
+ * <pre>{@code
+ * StreamsBuilder builder = new StreamsBuilder();
+ * KeyValueBytesStoreSupplier storeSupplier = Stores.inMemoryKeyValueStore("queryable-store-name");
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:55]: Why "queryable-store-name"?  IIRC we don't say "queryable store" anywhere else in the docs -- we use the term "interactive queries", if anything.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 11]
```diff
@@ -44,6 +44,40 @@
...
+ * For example, you can create an {@link org.apache.kafka.streams.kstream.Windowed windowed} RocksDB store with custom
+ * changelog configuration like:
+ * <pre>{@code
+ * Topology topology = new Topology();
+ * topology.addProcessor("processorName", ...);
+ *
+ * Map<String,String> changelogConfig = new HashMap<>();
+ * StoreBuilder<WindowStore<Integer, Long>> storeBuilder = Stores
+ *   .windowStoreBuilder(
+ *     Stores.persistentWindowStore("queryable-store-name", ...),
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: Why "queryable-store-name"?  IIRC we don't say "queryable store" anywhere else in the docs -- we use the term "interactive queries", if anything.

**Review by miguno — CHANGES_REQUESTED** [review]

File: streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java (hunk 4 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java:38]: I think plural is intended here, as we have "key-value", "time-windowed", and "session-windowed" RocksDB store. (Plural was there before... \cc @dguy what was your intention when doing the JavaDocs here?)

File: streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Materialized.java]: The key and value types `K` and `V` are not `Bytes`/`byte[]` but can be anything -- only the store `S` has a "nested" generic type that is `<Bytes,byte[]>`

I don't like the WS in `<Bytes,byte[]>` and keep as is.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 10 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:55]: We use parameter name `KStreamBuilder#table(String topic, String queryableStoreName)` in old API: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java#L400

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 12]
```diff
@@ -44,6 +44,40 @@
 
 /**
  * Factory for creating state stores in Kafka Streams.
+ * <p>
+ * When using the high-level DSL, i.e., {@link org.apache.kafka.streams.StreamsBuilder StreamsBuilder}, users create
+ * {@link StoreSupplier}s that can be further customized via
+ * {@link org.apache.kafka.streams.kstream.Materialized Materialized}.
+ * For example, a topic read as {@link org.apache.kafka.streams.kstream.KTable KTable} can be materialized into an
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:51]: is the second `KTable` intentional?

**Comment by bbejeck** [comment]
one minor question otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:51]: That is correct:

```
{@link org.apache.kafka.streams.kstream.KTable KTable}
```

will show on java docs as `org.apache.kafka.streams.kstream.KTable`, while the above will show as `KTable` whose ref links to `org.apache.kafka.streams.kstream.KTable` still.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 13]
```diff
@@ -44,6 +44,40 @@
...
+ * For example, you can create a {@link org.apache.kafka.streams.kstream.Windowed windowed} RocksDB store with custom
+ * changelog topic configuration like:
+ * <pre>{@code
+ * Topology topology = new Topology();
+ * topology.addProcessor("processorName", ...);
+ *
+ * Map<String,String> topicConfig = new HashMap<>();
+ * StoreBuilder<WindowStore<Integer, Long>> storeBuilder = Stores
+ *   .windowStoreBuilder(
+ *     Stores.persistentWindowStore("queryable-store-name", ...),
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:74]: nit: better be consistent with DSL examples to use `inMemoryKeyValueStore `?

**Review by guozhangwang — COMMENTED** [review]
One nit comment otherwise lgtm.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 13 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:74]: I intentionally put two different stores...

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:74]: Okay :) I guess it is personal taste then, I'm going to merge as is.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**